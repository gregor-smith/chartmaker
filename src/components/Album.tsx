import React, { FC } from 'react'
import { css, cx } from 'emotion'
import { Album as AlbumDetails, formatAlbumTitle } from '../state'


type Props = {
    details: AlbumDetails | null
    sizeRem: number
}


const baseStyle = css({
    backgroundColor: 'white',
    margin: '0.15rem'
})


export const Album: FC<Props> = ({ details, sizeRem }) => {
    const style = cx(
        baseStyle,
        css({
            width: `${sizeRem}rem`,
            height: `${sizeRem}rem`
        })
    )
    const image = details === null
        ? null
        : <img src={details.image.smallURL} alt={formatAlbumTitle(details)}/>
    return (
        <div className={style}>
            {image}
        </div>
    )
}
