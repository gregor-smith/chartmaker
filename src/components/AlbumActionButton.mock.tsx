const mock: typeof import('@/components/AlbumActionButton') = {
    AlbumActionButton: props =>
        <div {...props} className='mock-album-action-button'/>
}


export const AlbumActionButton = mock.AlbumActionButton
