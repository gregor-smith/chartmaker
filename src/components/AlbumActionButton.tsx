import { h, FunctionComponent } from 'preact'
import { css } from 'emotion'


type Props = {
    onClick: () => void
    title: string
}


const style = css({
    background: 'none',
    border: 'none'
})


export const AlbumActionButton: FunctionComponent<Props> = props =>
    <button {...props} class={style}/>
