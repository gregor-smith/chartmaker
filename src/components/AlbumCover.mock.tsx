import { cx } from 'emotion'


const mock: typeof import('@/components/AlbumCover') = {
    AlbumCover: ({
        album,
        size,
        overlayClass,
        children,
        highlighted,
        ...props
    }) => {
        const json = JSON.stringify(album)
        const className = cx('mock-album-cover', overlayClass)
        return (
            <div {...props} className={className}>
                {`Album: ${json}`}
                {`Size: ${size}`}
                {`Highlighted: ${highlighted}`}
                {children}
            </div>
        )
    }
}


export const AlbumCover = mock.AlbumCover
