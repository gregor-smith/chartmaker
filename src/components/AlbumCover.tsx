import React, { FC, DragEvent } from 'react'
import { css, cx } from 'emotion'

import { TEXT_COLOUR, ALBUM_PADDING_SIZE, highlightBackgroundStyle } from '@/style'
import { Album } from '@/types'


export type AlbumCoverProps = {
    album: Album
    size: string
    onDragStart: (event: DragEvent<HTMLDivElement>) => void
    onDragOver?: (event: DragEvent<HTMLDivElement>) => void
    onDragEnter?: (event: DragEvent<HTMLDivElement>) => void
    onDrop?: (event: DragEvent<HTMLDivElement>) => void
    onMouseEnter?: () => void
    overlayClass?: string
    highlighted?: boolean
}


const baseContainerStyle = css({
    position: 'relative',
    background: TEXT_COLOUR,
    margin: ALBUM_PADDING_SIZE
})

const imageStyle = css({
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    maxWidth: '100%',
    height: '100%',
    objectFit: 'contain'
})

// the overlay is given an additional non-emotion class so it can be used as a
// selector in the container's style
const overlaySelector = 'overlay'

const baseOverlayStyle = cx(
    overlaySelector,
    imageStyle,
    css({ zIndex: 10 })
)


export const AlbumCover: FC<AlbumCoverProps> = ({
    album,
    size,
    onDragStart,
    onDragEnter,
    onDragOver,
    onDrop,
    onMouseEnter,
    children,
    overlayClass,
    highlighted
}) => {
    const overlayStyle = cx(baseOverlayStyle, overlayClass)
    const containerStyle = cx(
        baseContainerStyle,
        css({
            width: size,
            height: size,
            [`:not(:hover) .${overlaySelector}`]: {
                display: 'none'
            }
        }),
        highlighted === false
            ? highlightBackgroundStyle
            : undefined
    )

    let image: JSX.Element | undefined
    if (!album.placeholder) {
        image = (
            // lazy loading causes problems with html2canvas so keep it default
            <img className={imageStyle}
                src={album.url}
                alt={album.name}
                crossOrigin='anonymous'/>
        )
    }

    return (
        <div className={containerStyle}
                draggable={!album.placeholder}
                onDragStart={onDragStart}
                onDragEnter={onDragEnter}
                onDragOver={onDragOver}
                onDrop={onDrop}
                onMouseEnter={onMouseEnter}
                title={album.placeholder ? undefined : album.name}>
            {image}
            <div className={overlayStyle}>
                {children}
            </div>
        </div>
    )
}
