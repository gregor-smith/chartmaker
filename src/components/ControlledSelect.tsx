import React, { FC, ChangeEvent } from 'react'

import { inputStyle } from '@/style'


export type ControlledSelectProps = {
    id?: string
    value: number
    onChange: (value: number) => void
}


export const ControlledSelect: FC<ControlledSelectProps> = ({ onChange, ...props }) => {
    function change(event: ChangeEvent<HTMLSelectElement>) {
        event.preventDefault()
        const value = Number(event.target.value)
        onChange(value)
    }

    return <select {...props} className={inputStyle} onChange={change}/>
}
