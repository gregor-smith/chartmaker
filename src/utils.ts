import html2canvas from 'html2canvas'

import { BACKGROUND_COLOUR } from './style'


export function readFileText(file: File): Promise<string> {
    return new Promise((resolve, reject) => {
        const reader = new FileReader()
        reader.addEventListener('error', reject)
        reader.addEventListener('load', () => resolve(reader.result as string))
        reader.readAsText(file, 'utf-8')
    })
}


export function findIndex<T>(array: ReadonlyArray<T>, predicate: (item: T) => boolean): number | null {
    const index = array.findIndex(predicate)
    return index === -1 ? null : index
}


export async function elementToDataURI(element: HTMLElement, scale: number) {
    const canvas = await html2canvas(element, {
        allowTaint: true,
        scale,
        useCORS: true,
        imageTimeout: 0,
        backgroundColor: BACKGROUND_COLOUR
    })
    return canvas.toDataURL()
}


export function jsonToDataURI(json: string): string {
    return 'data:application/json;charset=utf-8,' + encodeURIComponent(json)
}


export function downloadURI(uri: string, filename: string) {
    const link = document.createElement('a')
    link.href = uri
    link.download = filename
    link.click()
    link.remove()
}
