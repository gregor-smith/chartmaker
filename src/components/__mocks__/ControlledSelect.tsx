import React, { FC, ChangeEvent } from 'react'

import { ControlledSelectProps } from '@/components/ControlledSelect'


export const ControlledSelect: FC<ControlledSelectProps> = ({
    id,
    value,
    onChange,
    children
}) => {
    function change(event: ChangeEvent<HTMLDivElement>) {
        onChange(Number((event.target as any).value))
    }

    return (
        <div id={id} className='mock-controlled-select' onChange={change}>
            {`Value: ${value}`}
            {children}
        </div>
    )
}
