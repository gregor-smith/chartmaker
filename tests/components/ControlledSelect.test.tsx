import React from 'react'
import { render } from 'react-dom'
import { act } from 'react-dom/test-utils'

import { ControlledSelect } from '@/components/ControlledSelect'

import { RenderContainer, ignore, fireEvent } from '../utils'


const container = new RenderContainer()


test('renders select', () => {
    render(
        <ControlledSelect id='test-select'
                value={1}
                onChange={ignore}>
            Test children
        </ControlledSelect>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})


test('change event calls onChange prop', () => {
    const mock = jest.fn<void, [ number ]>()

    render(
        <ControlledSelect id='test-select'
                value={1}
                onChange={mock}>
            Test children
        </ControlledSelect>,
        container.element
    )

    act(() =>
        fireEvent(
            'change',
            container.element?.querySelector('select'),
            { target: { value: '2' } }
        )
    )

    expect(mock).toHaveBeenCalledTimes(1)
    expect(mock).toHaveBeenCalledWith(2)
})
