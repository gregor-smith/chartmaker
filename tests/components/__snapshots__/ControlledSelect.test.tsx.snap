// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders select 1`] = `
.emotion-0 {
  background: black;
  color: white;
  border: 1px solid white;
  padding: 0.25rem;
  width: 100%;
}

<div
  id="test-render-container"
>
  <select
    class="emotion-0"
    id="test-select"
  >
    Test children
  </select>
</div>
`;
